File: src/core/memorylane-sync.ts
import { execSync } from 'node:child_process'
import type { ContextLaneMemoryRecord } from './schema.js'

export interface SyncResult {
  saved: number
  failed: number
  errors: string[]
  method: 'http' | 'cli' | 'none'
}

export async function syncToMemoryLane(records: ContextLaneMemoryRecord[]): Promise<SyncResult> {
  const result: SyncResult = { saved: 0, failed: 0, errors: [], method: 'none' }

  const memoryLaneUrl = process.env.CONTEXTLANE_MEMORYLANE_URL

  if (memoryLaneUrl) {
    try {
      for (const record of records) {
        const res = await fetch(`${memoryLaneUrl}/v1/memorylane/memories`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(record),
        })
        if (res.ok) result.saved++
        else {
          result.failed++
          result.errors.push(`HTTP ${res.status}: ${await res.text()}`)
        }
      }
      result.method = 'http'
      return result
    } catch (e) {
      result.errors.push(`HTTP sync failed: ${e instanceof Error ? e.message : String(e)}`)
    }
  }

  try {
    execSync('which memorylane', { stdio: 'pipe' })
    for (const record of records) {
      try {
        const tags = record.tags.map(t => `--tag ${t}`).join(' ')
        execSync(`memorylane remember "${record.text.replace(/"/g, '\\"')}" ${tags} --tag contextlane`, { stdio: 'pipe', timeout: 10000 })
        result.saved++
      } catch {
        result.failed++
      }
    }
    result.method = 'cli'
    return result
  } catch {
    // memorylane CLI not available
  }

  result.method = 'none'
  result.errors.push('MemoryLane not available. Set CONTEXTLANE_MEMORYLANE_URL or install @talocode/memorylane.')
  return result
}

File: python/contextlane/core/memorylane_sync.py
import os
import subprocess
import urllib.request
import urllib.error
import json


async def sync_to_memory_lane(records: list) -> dict:
    result = {"saved": 0, "failed": 0, "errors": [], "method": "none"}

    url = os.environ.get("CONTEXTLANE_MEMORYLANE_URL")

    if url:
        try:
            for record in records:
                data = json.dumps(record).encode("utf-8")
                req = urllib.request.Request(
                    f"{url}/v1/memorylane/memories",
                    data=data,
                    headers={"Content-Type": "application/json"},
                )
                resp = urllib.request.urlopen(req, timeout=10)
                if resp.status == 200:
                    result["saved"] += 1
                else:
                    result["failed"] += 1
                    result["errors"].append(f"HTTP {resp.status}")
            result["method"] = "http"
            return result
        except Exception as e:
            result["errors"].append(f"HTTP sync failed: {e}")

    # CLI fallback
    try:
        subprocess.run(["which", "memorylane"], capture_output=True, check=True)
        for record in records:
            try:
                tags = " ".join(f"--tag {t}" for t in record.get("tags", []))
                text = record["text"].replace('"', '\\"')
                subprocess.run(
                    f'memorylane remember "{text}" {tags} --tag contextlane',
                    shell=True, capture_output=True, timeout=10,
                )
                result["saved"] += 1
            except Exception:
                result["failed"] += 1
        result["method"] = "cli"
        return result
    except FileNotFoundError:
        pass

    result["errors"].append("MemoryLane not available. Set CONTEXTLANE_MEMORYLANE_URL or install @talocode/memorylane.")
    return result
